src/flags.jl

```diff
@@ -0,0 +1,226 @@
+        (:PROPER_PAIR,   UInt16(0x002), "the read is mapped in a proper pair"                                       ),
+        (:UNMAP,         UInt16(0x004), "the read itself is unmapped; conflictive with FLAG_PROPER_PAIR"            ),
+        (:MUNMAP,        UInt16(0x008), "the mate is unmapped"                                                      ),
+        (:REVERSE,       UInt16(0x010), "the read is mapped to the reverse strand"                                  ),
+        (:MREVERSE,      UInt16(0x020), "the mate is mapped to the reverse strand"                                  ),
+        (:READ1,         UInt16(0x040), "this is read1"                                                             ),
+        (:READ2,         UInt16(0x080), "this is read2"                                                             ),
+        (:SECONDARY,     UInt16(0x100), "not primary alignment"                                                     ),
+        (:QCFAIL,        UInt16(0x200), "QC failure"                                                                ),
+        (:DUP,           UInt16(0x400), "optical or PCR duplicate"                                                  ),
+        (:SUPPLEMENTARY, UInt16(0x800), "supplementary alignment"                                                   ),]
+    @assert bits isa UInt16 "The bits must be of type UInt16."
+    sym = Symbol("FLAG_", name)
+    docstring = """    $sym
+    SAM/BAM flag: $doc
+
+    See also: [`flag`](@ref)
+    """
+    @eval begin
+        @doc $(docstring) const $(sym) = $(bits)
+    end
+end
+
+"""
+    ispaired(record::XAMRecord)::Bool
+
+Query whether the `record`'s template has multiple segments in sequencing.
+"""
+function ispaired(record::XAMRecord)::Bool
+    return flag(record) & FLAG_PAIRED == FLAG_PAIRED
+end
+
+"""
+    isproperpair(record::XAMRecord)::Bool
+
+Query whether each segment of the `record`'s template properly aligned according to the aligner.
+"""
+function isproperpair(record::XAMRecord)::Bool
+    return flag(record) & PROPER_PAIR == PROPER_PAIR
+end
+
+"""
+    isunmapped(record::XAMRecord)::Bool
+
+Query whether the `record` is unmapped.
+"""
+function isunmapped(record::XAMRecord)::Bool
+    return flag(record) & FLAG_UNMAP == FLAG_UNMAP
+end
+
+"""
+    ismapped(record::XAMRecord)::Bool
+
+Query whether the `record` is mapped.
+"""
+function ismapped(record::XAMRecord)::Bool
+    # return flag(record) & FLAG_UNMAP == 0
+    return isfilled(record) && (flag(record) & FLAG_UNMAP == 0)
+end
+
+"""
+    ismateunmapped(record::XAMRecord)::Bool
+
+Query whether the `record`'s mate is unmapped.
+"""
+function ismateunmapped(record::XAMRecord)::Bool
+    return flag(record) & FLAG_MUNMAP == FLAG_MUNMAP
+end
+
+"""
+    isnextmapped(record::XAMRecord)::Bool
+
+Test if the mate/next read of `record` is mapped.
+"""
+function isnextmapped(record::XAMRecord)::Bool
+    return flag(record) & FLAG_MUNMAP == 0
```

Comment:
Why is this called "next", where the opposite one is called "mate"?

Reply:
This is probably a hangover from the original XAM. Let us adjust to the more holistic template-segment terminology.

Reply:
I've used "next" instead of "mate".